processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java

```diff
@@ -125,11 +125,11 @@ public <T> T as(Class<T> clazz)
       return (T) WrappedTimeBoundaryInspector.create(baseSegment.as(TimeBoundaryInspector.class));
     } else if (TopNOptimizationInspector.class.equals(clazz)) {
       // if the baseFilter is not null, then rows from underlying cursor can be potentially filtered.
-      // otherwise, a filtering inner join can also filter rows.
+      // otherwise, a filtering inner or left join can also filter rows.
       return (T) new SimpleTopNOptimizationInspector(
-          baseFilter != null || clauses.stream().anyMatch(
-              clause -> clause.getJoinType() == JoinType.INNER && !clause.getCondition().isAlwaysTrue()
-          )
+          !(baseFilter != null || clauses.stream().anyMatch(
+              clause -> clause.getJoinType().isLefty() && !clause.getCondition().isAlwaysTrue()
```

Comment:
I think this isn't correct. The condition we want is that the `baseFilter` is null, and all clauses are either lefty or always-true. Like this:

```
baseFilter == null && clauses.stream.allMatch(
  clause -> clause.getJoinType().isLefty() || clause.getCondition().isAlwaysTrue()
)
```

Reply:
oops, yes, fixed again. I copied your suggestion wrong last time (lost the ! on lefty check) and then flipped it to match the new contract of the inspector thingy, but should have just rewritten it since this is clearer.